errors/errors.go

```diff
@@ -18,82 +18,136 @@ package errors
+// must map to a format which is a string containing ${formatArgu1}.
+// Exp:
+// Reason "kind:NotFound" may map to Format "${kindName} was not found".
+// Reason "Status:Sleep" may map to Format "${Name} is sleeping now"
 type Reason string
 
+// message was the result returned to client.
+// Exp:
+// {
+//   "message": "name japari is too short",
+//   "reason": "monitoring:CreateDashboardNameTooShort",
+//   "data": {
+//     "name": "japari"
+//   }
+// }
 type message struct {
-	// Required, a machine-readable description of the error.
-	Reason Reason `json:"reason"`
-	// Required when template is used in message or i18nMessage.
-	Data []interface{} `json:"data,omitempty"`
-	// Required for 4xx but optional for 5xx. Message is a human-readable description
-	// of the error. Message can be golang template.
-	Message string `json:"message"`
+	Reason  Reason            `json:"reason,omitempty"`
+	Message string            `json:"message"`
+	Data    map[string]string `json:"data,omitempty"`
 }
 
-// err is an error intended to be used by all APIs to return error to clients.
 type err struct {
-	// Suggested HTTP return code for this error, 0 if not set. This field is optional.
-	// Caller can choose to use this code or choose to use another error code for client.
-	code int
-	// Format is used to generate message.
+	message
+	code   int
 	format string
-	// Useful message.
-	message message
 }
 
 func (e *err) Code() int {
 	return e.code
 }
 
 func (e *err) Message() interface{} {
-	return &e.message
-}
-
-func (e *err) Reason() Reason {
-	return e.message.Reason
+	return e.message
 }
 
 func (e *err) Error() string {
 	return e.message.Message
 }
 
-type formatter struct {
+// Factory is an error factory.
+type Factory struct {
 	code   int
 	reason Reason
-	// It should like
-	// 1. something named {0} is not found
-	// 2. something named {name} is not found
-	// Which one will win the battle?
 	format string
 }
 
-func (f *formatter) Format(a ...interface{}) Error {
+// New generates an error.
+func (f *Factory) New(v ...interface{}) error {
+	msg := message{Reason: f.reason}
+	msg.Message, msg.Data = expand(f.format, v...)
 	return &err{
-		code:   f.code,
-		format: f.format,
-		message: message{
-			Reason:  f.reason,
-			Data:    a,
-			Message: fmt.Sprintf(f.format, a...),
-		},
+		message: msg,
+		code:    f.code,
+		format:  f.format,
 	}
 }
 
-func ResourceNotFound(resource string) Error {
-	return (&formatter{
-		code:   http.StatusNotFound,
-		reason: Reason(http.StatusText(http.StatusNotFound)),
-		format: "%s is not found",
-	}).Format(resource)
+// CanNew checks whether f is able to New an error which has the same code, reason and format with e.
+func (f *Factory) CanNew(e error) bool {
+	x, ok := e.(*err)
+	if !ok {
+		return false
+	}
+	return f.code == x.code && f.reason == x.message.Reason && f.format == x.format
+}
+
+// Type maps to http code.
+// And it can be used to make an error factory.
+type Type struct {
```

Comment:
May StatusCode is more appropriate。

Reply:
Different status code means different type of errors, so I used Type here.